### PR TITLE
#260. Implement logout.

### PR DESCRIPTION
**Pull request by ivan-karavaiev**

### CODE REVIEW

File: ios/Codeforces Watcher/Application/features/auth/signIn/SignInViewController.swift [hunk 1]
```diff
@@ -44,6 +44,7 @@ class SignInViewController: ClosableViewController, ReKampStoreSubscriber {
         switch (state.signInStatus) {
         case .done:
             hideLoading()
+            store.dispatch(action: AuthRequests.DestroyStatus())
             closeViewController()
```
Comment (MatyashDen) [inline ios/Codeforces Watcher/Application/features/auth/signIn/SignInViewController.swift:48]: I suppose it's also needed in sign up screen(in android also)

File: ios/Codeforces Watcher/Application/features/users/UserViewController.swift [hunk 2]
```diff
@@ -54,15 +54,26 @@ class UserViewController: ClosableViewController, ReKampStoreSubscriber {
...
         view.backgroundColor = Palette.white
         
+        setupRightBarButton()
+        
+        buildViewTree()
+        setConstraints()
+    }
+    
+    private func setupRightBarButton() {
+        let isUserAccount = (store.state.users.userAccount?.codeforcesUser?.handle == handle)
```
Comment (MatyashDen) [inline ios/Codeforces Watcher/Application/features/users/UserViewController.swift:64]: not always true, it's better to pass this to constructor

File: ios/Codeforces Watcher/Application/features/users/UserViewController.swift [hunk 3]
```diff
@@ -54,15 +54,26 @@ class UserViewController: ClosableViewController, ReKampStoreSubscriber {
...
+            action: isUserAccount ? #selector(doLogOut) :  #selector(showDeleteUserAlert)
         )
-        
-        buildViewTree()
-        setConstraints()
+    }
+    
+    @objc private func doLogOut() {
+        store.dispatch(action: AuthRequests.LogOut())
+        presentingViewController?.dismiss(animated: true)
```
Comment (MatyashDen) [inline ios/Codeforces Watcher/Application/features/users/UserViewController.swift:76]: Could you add alert before logging out(on Android also)?

File: ios/Codeforces Watcher/Application/features/users/cells/user/UserTableViewCell.swift [hunk 4]
```diff
@@ -63,47 +63,73 @@ class UserTableViewCell: UITableViewCell {
...
+        handleLabel.attributedText = user.handleText
+        ratingLabel.attributedText = user.ratingText
+        dateOfLastRatingUpdateLabel.text = user.ratingUpdateDateText
+        valueOfLastRatingUpdateLabel.attributedText = user.ratingUpdateValueText
+    }
+}
 
-        if let ratingChange = user.ratingChanges.last {
+fileprivate extension UserItem.UserItem {
+
```
Comment (MatyashDen) [inline ios/Codeforces Watcher/Application/features/users/cells/user/UserTableViewCell.swift:99]: Nice refactoring 👍

**Review by MatyashDen — CHANGES_REQUESTED** [review]

File: ios/Codeforces Watcher/Application/features/auth/signIn/SignInViewController.swift (hunk 1 above)
Comment (MatyashDen) [reply ios/Codeforces Watcher/Application/features/auth/signIn/SignInViewController.swift:48]: well, you can do it in middleware on LogOut.Success

File: ios/Codeforces Watcher/Application/features/users/UserViewController.swift [hunk 5]
```diff
@@ -67,13 +67,27 @@ class UserViewController: ClosableViewController, ReKampStoreSubscriber {
...
+            message: "log_out_ask".localized,
+            preferredStyle: .alert
+        )
+        
+        let logOutButton = UIAlertAction(title: "log_out".localized.uppercased(), style: .destructive) {_ in
+            store.dispatch(action: AuthRequests.LogOut())
+            self.presentingViewController?.dismiss(animated: true)
+        }
+        let stayLoggedInButton = UIAlertAction(title: "stay_logged_in".localized.uppercased(), style: .cancel)
+
```
Comment (MatyashDen) [inline ios/Codeforces Watcher/Application/features/users/UserViewController.swift]: Let's remove uppercase here and in other alert

**Review by MatyashDen — APPROVED** [review]
Ship it!

File: ios/Codeforces Watcher/Application/features/users/UserViewController.swift (hunk 5 above)
Comment (ivan-karavaiev) [reply ios/Codeforces Watcher/Application/features/users/UserViewController.swift]: Why?

**Pull request merged**